### ISSUE TITLE
Draw adjudication

### ISSUE THREAD

**Issue by iopq** [lightvector/KataGo#191]
I would like KataGo to output draw probabilities in order to finish drawn games with integer komi faster (90% draw chance for 3 moves in a row from both sides). It might not be important for 19x19, but like half of the games on 9x9 are draws with 7 komi

**Comment by lightvector** [comment]
I'm basically not going to do this myself. When I last attempted a win-loss-draw head to also output draw probabilities, as I discussed a little in the Leela Zero discord, the result was basically a disaster. Either it was buggy (but I'm not sure how) or for whatever reason it caused the early training to massively overfit, or it greatly changed what the correct scale on of what the value head loss needed to be relative to the policy loss, or something.

Part of why it may have caused overfitting was precisely because of the sawtooth non-monotonicity of whether a draw is usually-possible at even/odd komi under area rules, which *also* depends jointly on the parity of the board size. My understanding is that highly-XOR-like features in a problem are typically thought to make fitting more difficult for neural nets, and a draw head exacerbates this because the output of the neural net in logit space now needs to vary massively depending on this parity, rather than varying only slightly. Some of this was also reflected in the winrate now being oscillatory and non-monotone in the komi.

Or, again my implementation could have been buggy. :)

Either way, I do not plan to make a second attempt at it. But I also have no objections to if you or someone else wants to do it. If you can demonstrate an implementation such that learning is not harmed when comparing between otherwise controlled two-side-by-side runs up to at least human pro level on 19x19, with this being the only difference, and such that there is no significant runtime cost either, then I would consider it. You're also welcome to write a branch with such a change as just an alternate fork without such testing, if you don't care about ever having your code merged in.

**Comment by isty2e** [comment]
Presumably adding a +0.5 komi input feature (1 if +0.5 and 0 if integer or the other way) with a WDL value head might help keeping things continuous? I'm not particularly interested in testing it, though.

**Comment by iopq** [comment]
Well, at first I just thought it would be NICE to have because I'm specifically looking for "perfect" drawn games on 9x9. But I've now identified an actual training benefit from this. I might take a shot at it.

**Comment by iopq** [comment]
I figured how you could do draws. You output a probability of a "close game". A close game is a draw with integer komi rounded to odd values on board sizes with an odd amount of interesection and even values on board sizes with an even amount of interesections
In other words, with 7.5 komi on a 19x19 board you would say "there's a 7% chance that the lead without komi is going to be 7" and add that to white win rate since white wins close games
with 6.5 komi it would not change, except you add that to the black win rate
with 7.0 komi you would add it to the draw winrate
it should work equally well with Japanese and Chinese counting, I'm assuming KG does the score properly for both now

So where do I look to start adding an output to train on?
